### PR TITLE
[Distributed] make finalizer messages threadsafe

### LINKED ISSUE #42126: [Distributed] error in running finalizer: ConcurrencyViolationError(msg="lock must be held") (any_gc_flag)
This is a concurrency violation on this condition https://github.com/JuliaLang/julia/blob/8812c5cc242435bf267285fcf3e41e632c84e52c/stdlib/Distributed/src/remotecall.jl#L259

Happens when running tests on https://github.com/krynju/Dagger.jl/commit/a5b5267e1db1592164d3dbe2700dbe32625f10db (DTable groupby tests with Dagger on processes)

Windows, Julia master

Seems like the notify needs to be wrapped in a while->trylock loop (because it's running in finalizer context), since the lock on that condition doesn't seem to be obtained anywhere and the docs say that the Condition structure is not threadsafe

Will look more into that and open a PR sometime next week

```julia
From worker 2:    error in running finalizer: ConcurrencyViolationError(msg="lock must be held")
From worker 2:    concurrency_violation at .\condition.jl:8
From worker 2:    assert_havelock at .\condition.jl:25 [inlined]
From worker 2:    assert_havelock at .\condition.jl:48 [inlined]
From worker 2:    assert_havelock at .\condition.jl:72 [inlined]
From worker 2:    notify at .\condition.jl:132
From worker 2:    #notify#564 at .\condition.jl:130 [inlined]
From worker 2:    notify at .\condition.jl:130 [inlined]
From worker 2:    notify at .\condition.jl:130 [inlined]
From worker 2:    process_worker at C:\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v1.8\Distributed\src\remotecall.jl:288 [inlined] 
From worker 2:    send_del_client_no_lock at C:\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v1.8\Distributed\src\remotecall.jl:280  
From worker 2:    finalize_ref at C:\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v1.8\Distributed\src\remotecall.jl:94
From worker 2:    jl_apply at /cygdrive/c/buildbot/worker/package_win64/build/src\julia.h:1771 [inlined]
From worker 2:    run_finalizer at /cygdrive/c/buildbot/worker/package_win64/build/src\gc.c:278
From worker 2:    jl_gc_run_finalizers_in_list at /cygdrive/c/buildbot/worker/package_win64/build/src\gc.c:365
From worker 2:    run_finalizers at /cygdrive/c/buildbot/worker/package_win64/build/src\gc.c:394 [inlined]
From worker 2:    run_finalizers at /cygdrive/c/buildbot/worker/package_win64/build/src\gc.c:372
From worker 2:    jl_gc_run_pending_finalizers at /cygdrive/c/buildbot/worker/package_win64/build/src\gc.c:405
From worker 2:    jl_mutex_unlock at /cygdrive/c/buildbot/worker/package_win64/build/src\locks.h:133 [inlined]
From worker 2:    jl_generate_fptr at /cygdrive/c/buildbot/worker/package_win64/build/src\jitlayers.cpp:357
From worker 2:    jl_compile_method_internal at /cygdrive/c/buildbot/worker/package_win64/build/src\gf.c:1978
From worker 2:    jl_compile_method_internal at /cygdrive/c/buildbot/worker/package_win64/build/src\gf.c:1932 [inlined]
From worker 2:    _jl_invoke at /cygdrive/c/buildbot/worker/package_win64/build/src\gf.c:2237 [inlined]
From worker 2:    jl_apply_generic at /cygdrive/c/buildbot/worker/package_win64/build/src\gf.c:2427
From worker 2:    jl_apply at /cygdrive/c/buildbot/worker/package_win64/build/src\julia.h:1771 [inlined]
From worker 2:    start_task at /cygdrive/c/buildbot/worker/package_win64/build/src\task.c:881
```

### PR DESCRIPTION
**Pull request by vchuravy**
Partial revert of #41722

Fixes #42126

### CODE REVIEW

**Comment by Sacha0** [comment]
Tentatively marking backport-1.7, given the issue this patch targets has been observed on 1.7 previously.

**Review by vtjnash — DISMISSED** [review]
From CI, this hangs

**Comment by vtjnash** [comment]
Looks like this would be vulnerable to a lock priority inversion deadlock situation

File: stdlib/Distributed/src/messages.jl [hunk 1]
```diff
@@ -126,20 +126,21 @@ function flush_gc_msgs(w::Worker)
     if !isdefined(w, :w_stream)
         return
     end
-    lock(w.msg_lock) do
-        w.gcflag || return # early exit if someone else got to this
+    local add_msgs::Vector{Any}
+    local del_msgs::Vector{Any}
+    lock(w.msg_lock) do # TODO: lock_inversion?
```
Comment (vchuravy) [inline stdlib/Distributed/src/messages.jl:131]:
```suggestion
    lock(w.msg_lock) do
```

File: stdlib/Distributed/src/messages.jl [hunk 2]
```diff
@@ -126,21 +126,20 @@ function flush_gc_msgs(w::Worker)
...
-
-    # del_msgs gets populated by finalizers, so be very careful here about ordering of allocations
-    # XXX: threading requires this to be atomic
-    new_array = Any[]
-    msgs = w.del_msgs
-    w.del_msgs = new_array
-    if !isempty(msgs)
-        #print("sending delete of $msgs\n")
+    if !isempty(del_msgs)
         remote_do(del_clients, w, msgs)
```
Comment (vchuravy) [inline stdlib/Distributed/src/messages.jl:143]:
```suggestion
        remote_do(del_clients, w, del_msgs)
```

**Comment by tkf** [comment]
Would using a lock-free queue, stack, or "channel" for the messages help here? It sounds nice to not block and do minimal work inside a finalizer.

**Comment by vchuravy** [comment]
> Would using a lock-free queue, stack, or "channel" for the messages help here? It sounds nice to not block and do minimal work inside a finalizer. (You can always `schedule` a task and so maybe this is not the issue/solution though.)

Yeah that would be a good optimization, but I don't think we have that in base right now?

But even with a lock-free queue, notifying the consumer is a problem.

**Comment by tkf** [comment]
For a queue with notification, we can use a lock-free unbounded "channel" (aka dual queue). It's possible to implement lock-free(-ish) `push!` even when `popfirst!` can wait. One catch is that `push!` needs to be able to run `schedule` [^1] but it's already usable in the finalizer. I think I can create a PR for adding the queue in this: Scherer & Scott (2004) Nonblocking Concurrent Data Structures with Condition Synchronization https://link.springer.com/chapter/10.1007/978-3-540-30186-8_13

[^1]: It's also conceivable to use lock-free(-ish) queues inside the Julia runtime. It'd then make `push!` "more" lock-free.

**Comment by vchuravy** [comment]
@KristofferC I would like to let this bake on master for a while before backporting. Should I remove the tag until a week or two from now?

**Comment by KristofferC** [comment]
Sounds good.

**Pull request merged**